File: Problems_and_Solutions/days-20 to 29/day-22/Java/Solution1.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

class Solution1 {
    public static ListNode oddEvenList(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }
        
        ListNode odd = head;
        ListNode even = head.next;
        ListNode evenHead = even;

        while(even != null && even.next != null){
            odd.next = odd.next.next;
            even.next = even.next.next;

            odd = odd.next;
            even = even.next;
        }
        odd.next = evenHead;
        return head;
    }

    public static void printList(ListNode head) {
        while(head != null){
            System.out.print(head.val + " ");
            head = head.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);

        printList(head);
        ListNode res = oddEvenList(head);
        printList(res);
    }
}

File: Problems_and_Solutions/days-20 to 29/day-22/c++/solution 1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
      if(head==NULL)
        return head;
      ListNode* oddPointer=head;
      ListNode* temp=head->next;
      ListNode* evenPointer=head->next;
      while(oddPointer->next!=NULL && evenPointer->next!=NULL)
      {
        oddPointer->next=evenPointer->next;
        oddPointer=oddPointer->next;
        
        evenPointer->next=oddPointer->next;
        evenPointer=evenPointer->next;
      }
      oddPointer->next=temp;
    return head;      
    }
};